idlparser/src/ffi/ast/visitor.rs

```diff
@@ -1,292 +1,479 @@
 use super::*;
 use crate::{ast as raw_ast, ast::visitor as raw_visitor, ast::visitor::Visitor as RawVisitor};
 use std::ptr;
+use wrapper::VisitorWrapper;
 
 #[repr(C, packed)]
 pub struct Visitor {
-    visit_service: extern "C" fn(*mut Visitor, *const Service),
-    visit_type: extern "C" fn(*mut Visitor, *const Type),
-    visit_optional_type_decl: extern "C" fn(*mut Visitor, *const TypeDecl),
-    visit_vector_type_decl: extern "C" fn(*mut Visitor, *const TypeDecl),
-    visit_result_type_decl: extern "C" fn(*mut Visitor, *const TypeDecl, *const TypeDecl),
-    visit_primitive_type_id: extern "C" fn(*mut Visitor, *const PrimitiveType),
-    visit_user_defined_type_id: extern "C" fn(*mut Visitor, *const u8, u32),
-    visit_func: extern "C" fn(*mut Visitor, *const Func),
-    visit_func_param: extern "C" fn(*mut Visitor, *const FuncParam),
-    visit_func_output: extern "C" fn(*mut Visitor, *const TypeDecl),
-    visit_struct_def: extern "C" fn(*mut Visitor, *const StructDef),
-    visit_struct_field: extern "C" fn(*mut Visitor, *const StructField),
-    visit_enum_def: extern "C" fn(*mut Visitor, *const EnumDef),
-    visit_enum_variant: extern "C" fn(*mut Visitor, *const EnumVariant),
-}
-
-/// # Safety
-///
-/// See documentation for [`const_ptr::as_ref`] and [`mut_ptr::as_mut`]
+    visit_service: unsafe extern "C" fn(context: *const (), *const Service),
```

Comment:
Can we use `Ptr` struct in this file too? Never mind if I missed the point

Reply:
We can, but it doesn't matter much as this Ptr is never touched by the parser